torch/csrc/lazy/core/lazy_mode.cpp

```diff
@@ -0,0 +1,75 @@
+#include <torch/csrc/lazy/core/lazy_mode.h>
+
+#include <c10/core/DispatchKey.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+#include <torch/csrc/lazy/backend/backend_device.h>
+#include <torch/csrc/lazy/core/lazy_graph_executor.h>
+#include <torch/library.h>
+
+// TODO(whc) we can't actually depend on ts backend code from here, but we could refactor if reusing the callback
+// turns out to be the best way to implement this
+#include <torch/csrc/lazy/ts_backend/ts_eager_fallback.h>
+
+namespace torch {
+namespace lazy {
+
+std::atomic<bool>& in_lazy_mode() {
+    static std::atomic<bool> _in_lazy_mode{false};
+    return _in_lazy_mode;
+}
+
+void LazyModeEnter(c10::Device device) {
+    in_lazy_mode() = true;
+    // It is straightforward why we want to set the lazy key on entering the mode:
+    // we force operators (even on regular eager tensors) to route to lazy implementations
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Lazy, true);
+}
+
+void LazyModeExit(c10::Device device) {
+    // Equally straightforward is that we no longer want the lazy key when we exit the mode:
+    // this lets operations on eager tensors outside the mode go back to normal eager behavior
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Lazy, false);
+
+    // Less obvious is that we also set an 'unlazy' key on mode exit, which lets us specially handle
+    // any 'lazy' tensors that are alive after the mode exit.  This could be avoided if we can find another way
+    // to make lazy tensors interoperable with eager kernels.
+    // TODO(whc) PrivateUse1 is just for prototyping
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::PrivateUse1, true);
```

Comment:
This turns on DispatchKey::PrivateUse1 for ALL future operations, even if the future operations do not involve any Lazy Tensors that were created inside of Lazy Mode.

What we probably want here is to put PrivateUse1 into the keyset of all the live tensors. To accomplish this we'd add a method to TensorImpl called `add_unlazy_key()` and it just adds the key to the key set. Something like https://github.com/pytorch/pytorch/blob/3eea311c0c1121db103d088037525be433380947/c10/core/TensorImpl.h#L965-L967

Reply:
Nice~ This sounds like a better solution in the long run.

If we have a lazy exit function where we can properly modify dispatch keys for tensors, we might as well consider `convert` them into eager compatible tensors as well, for two reasons: 1. This would be closer to what we have planned as the long-term solution to our hybrid method three; 2. the benefit of having an `unlazy` dispatch is that we don't have to tinker with `live` tensor at exit. But if we are already doing that, adding `unlazy` dispatch seems to be adding little value.

Having said that, just to move things quicker, we might also just blindly add `unlazy` to all tensors and pay the overhead to save some prototyping time. Later down the road we might have a better chance to choose the right implementation.

Reply:
> if we have a lazy exit function where we can properly modify dispatch keys for tensors

We definitely can modify dispatch keys of lazy tensors at exit.

> we might as well consider convert them into eager compatible tensors as well

Swapping the dispatch key (removing 'lazy') on a LTCTensorImpl just isn't enough.  Our 'proposal 2' tried to actually change to a new TensorImpl, but we ran into trouble with where to update the pointers to our old LTCTensorImpl.  

Luckily, I think Richard (or someone) came up with a nice suggestion that we didn't consider in our 3 proposals.  Basically, we add new fields (and logic) to LTCTensorImpl to make it so it can behave properly as an eager tensor when we ask it to.  This is kind of the inverse of our proposal 1, which is to modify the normal TensorImpl to add a field for a 'lazy metadata'.

We can start by doing the unlazy thing with Richard's fixes using `TESTING_ONLY_GenericWrapper` key and setting it only on 'live tensors' rather than all tensors- this should be quick.  Then if things look good, we could try the next step of making 'unlazy' unnecessary by adding 'eager' fields to LTCTensorImpl.

Reply:
> Luckily, I think Richard (or someone) came up with a nice suggestion that we didn't consider in our 3 proposals. Basically, we add new fields (and logic) to LTCTensorImpl to make it so it can behave properly as an eager tensor when we ask it to. This is kind of the inverse of our proposal 1, which is to modify the normal TensorImpl to add a field for a 'lazy metadata'.

Yes it's Richard's idea and that's what I was suggesting. Sorry I didn't make it clear in my earlier post.

> We can start by doing the unlazy thing with Richard's fixes using TESTING_ONLY_GenericWrapper key and setting it only on 'live tensors' rather than all tensors- this should be quick. Then if things look good, we could try the next step of making 'unlazy' unnecessary by adding 'eager' fields to LTCTensorImpl.

^^^ sounds like a good plan moving forward!